### PR TITLE
inventory: vultr: implement keyed groups and filter

### PR DESCRIPTION
**Pull request by resmo**
##### SUMMARY
implement keyed groups and filter. This supersedes #53165. Closes #53165 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
vultr

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/plugins/inventory/vultr.py [hunk 1]
```diff
@@ -121,9 +153,17 @@ def _retrieve_servers(api_key):
         raise AnsibleError("Error while fetching %s: %s" % (api_url, to_native(e)))
 
 
-class InventoryModule(BaseInventoryPlugin):
+class InventoryModule(BaseInventoryPlugin, Constructable):
+
     NAME = 'vultr'
 
+    def verify_file(self, path):
```
Comment (Spredzy) [inline lib/ansible/plugins/inventory/vultr.py:164]: Any reason to enforce that ? I know it's been used in other inventory plugins, but shouldn't we let user use the file name they want without enforcing it ? Plus most plugins do not specify in their documentation what the name should look like, leading to few mins of troubleshooting before "grasping" it should follow a specifal pattern

File: lib/ansible/plugins/inventory/vultr.py [hunk 2]
```diff
@@ -135,14 +175,32 @@ def parse(self, inventory, loader, path, cache=True):
             raise AnsibleError('Could not find an API key. Check inventory file and Vultr configuration files.')
 
         hostname_preference = self.get_option('hostname')
-        for server in _retrieve_servers(api_key):
+
+        # Add a top group 'vultr'
+        self.inventory.add_group(group='vultr')
```
Comment (Spredzy) [inline lib/ansible/plugins/inventory/vultr.py:184]: :+1:

**Review by Spredzy — APPROVED** [review]
Small nit but LGTM otherwise

File: lib/ansible/plugins/inventory/vultr.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/plugins/inventory/vultr.py:164]: considering the the problems 'unspecified' brings to the table, i would insist, yes

all other plugins should have this clearly documented, if not, its a bug in that plugin.

File: lib/ansible/plugins/inventory/vultr.py [hunk 3]
```diff
@@ -9,52 +9,82 @@
...
-      - "Yanis Guenane (@Spredzy)"
+        - "Yanis Guenane (@Spredzy)"
     short_description: Vultr inventory source
+    version_added: "2.7"
+    extends_documentation_fragment:
+        - constructed
     description:
         - Get inventory hosts from Vultr public cloud.
-        - Uses C(api_config), C(~/.vultr.ini), C(./vultr.ini) or VULTR_API_CONFIG path to config file.
+        - Uses I(api_config), I(~/.vultr.ini), I(./vultr.ini) or C(VULTR_API_CONFIG) path to config file.
```
Comment (bcoca) [inline lib/ansible/plugins/inventory/vultr.py]: inventory source file naming requirements?

File: lib/ansible/plugins/inventory/vultr.py [hunk 4]
```diff
@@ -9,52 +9,82 @@
...
         api_key:
             description: Vultr API key. If not specified will be taken from regular Vultr configuration.
+            type: str
             env:
-              - name: VULTR_API_KEY
+                - name: VULTR_API_KEY
         hostname:
             description: Field to match the hostname. Note v4_main_ip corresponds to the main_ip field returned from the API and name to label.
-            type: string
+            type: str
```
Comment (bcoca) [inline lib/ansible/plugins/inventory/vultr.py]: both forms are valid ... string is more descriptive to users (non programmers)

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/inventory/vultr.py (hunk 4 above)
Comment (resmo) [reply lib/ansible/plugins/inventory/vultr.py]: ok then. reverting

**Comment by resmo** [comment]
Thanks @bcoca and @Spredzy for the review!

**Review by bcoca — APPROVED** [review]

**Pull request merged**